drivers/metal/rendering_device_driver_metal.mm

```diff
@@ -0,0 +1,3919 @@
+
+	return OK;
+}
+
+RDD::FramebufferID RenderingDeviceDriverMetal::swap_chain_acquire_framebuffer(CommandQueueID p_cmd_queue, SwapChainID p_swap_chain, bool &r_resize_required) {
+	DEV_ASSERT(p_cmd_queue.id != 0);
+	DEV_ASSERT(p_swap_chain.id != 0);
+
+	SwapChain *swap_chain = (SwapChain *)(p_swap_chain.id);
+	if (context_driver->surface_get_needs_resize(swap_chain->surface)) {
+		r_resize_required = true;
+		return FramebufferID();
+	}
+
+	if (swap_chain->frame_buffer.drawable == nil) {
+		id<CAMetalDrawable> drawable = swap_chain->layer.nextDrawable;
+		ERR_FAIL_NULL_V_MSG(drawable, RDD::FramebufferID(), "no drawable available");
+		CGSize size = swap_chain->layer.drawableSize;
+		swap_chain->frame_buffer.set_drawable_and_size(drawable, Size2i(size.width, size.height));
+	}
+
+	return RDD::FramebufferID(&swap_chain->frame_buffer);
+}
+
+RDD::RenderPassID RenderingDeviceDriverMetal::swap_chain_get_render_pass(SwapChainID p_swap_chain) {
+	const SwapChain *swap_chain = (const SwapChain *)(p_swap_chain.id);
+	return swap_chain->render_pass;
+}
+
+RDD::DataFormat RenderingDeviceDriverMetal::swap_chain_get_format(SwapChainID p_swap_chain) {
+	const SwapChain *swap_chain = (const SwapChain *)(p_swap_chain.id);
+	return swap_chain->data_format;
+}
+
+void RenderingDeviceDriverMetal::swap_chain_free(SwapChainID p_swap_chain) {
+	SwapChain *swap_chain = (SwapChain *)(p_swap_chain.id);
+	_swap_chain_release(swap_chain);
+	render_pass_free(swap_chain->render_pass);
+	memdelete(swap_chain);
+}
+
+#pragma mark - Frame buffer
+
+RDD::FramebufferID RenderingDeviceDriverMetal::framebuffer_create(RenderPassID p_render_pass, VectorView<TextureID> p_attachments, uint32_t p_width, uint32_t p_height) {
+	MDRenderPass *pass = (MDRenderPass *)(p_render_pass.id);
+
+	Vector<MTL::Texture> textures;
+	textures.resize(p_attachments.size());
+
+	for (uint32_t i = 0; i < p_attachments.size(); i += 1) {
+		MDAttachment const &a = pass->attachments[i];
+		id<MTLTexture> tex = rid::get(p_attachments[i]);
+		if (tex == nil) {
+#if DEV_ENABLED
+			WARN_PRINT("Invalid texture for attachment " + itos(i));
+#endif
+		}
+		if (a.samples > 1) {
+			if (tex.sampleCount != a.samples) {
+#if DEV_ENABLED
+				WARN_PRINT("Mismatched sample count for attachment " + itos(i) + "; expected " + itos(a.samples) + ", got " + itos(tex.sampleCount));
+#endif
+			}
+		}
+		textures.write[i] = tex;
+	}
+
+	MDFrameBuffer *fb = new MDFrameBuffer(textures, Size2i(p_width, p_height));
+	return FramebufferID(fb);
+}
+
+void RenderingDeviceDriverMetal::framebuffer_free(FramebufferID p_framebuffer) {
+	MDFrameBuffer *obj = (MDFrameBuffer *)(p_framebuffer.id);
+	delete obj;
+}
+
+#pragma mark - Shader
+
+const uint32_t SHADER_BINARY_VERSION = 1;
+
+// region Serialisation
+
+class BufWriter;
+
+template <typename T>
+concept Serializable = requires(T t, BufWriter &p_writer) {
+	{ t.serialize_size() } -> std::same_as<size_t>;
+	{ t.serialize(p_writer) } -> std::same_as<void>;
+};
```

Comment:
I don't think we're letting concepts into the codebase at this point. Could you use `static_assert` instead?

Reply:
Indeed, Godot still targets C++17 for the foreseeable future, and concepts are a C++20 feature.

Reply:
I do understand that for all the shared, cross-platform library code, C++ 17 is the standard. 

If I may plead my case 😊, my reasoning for this is that the Objective-C code is Apple (macOS, iOS, tvOS, xrOS) specific, and will have access to C++ 20 features. I haven't used them extensively, but it did make this code simple to express and write. 

I will explore the `static_assert` option.

Reply:
I tried `static_assert`, but the code was more complicated and harder to follow. Given this is technically Objective-C++ code, could we pretend this is an Objective-C++ feature 😊.

Reply:
On my part, I can accept that. 😀 Thanks for trying anyway.

Reply:
Thank you!

I will of course make all the other changes you requested.